File: 백준/Gold/7569. 토마토/토마토.py
import sys
from collections import deque
fastinput = sys.stdin.readline
M, N, H = map(int, fastinput().split())
box = [[] for i in range(H)]
for i in range(H):
    for j in range(N):
        box[i].append(list(map(int, input().split())))
q = deque()
d_end = 0
for i in range(len(box)):
    for j in range(len(box[i])):
        for k in range(len(box[i][j])):
            if box[i][j][k] == 1:
                q.append((i,j,k))
                d_end += 1
dx = [0,0,-1,1,0,0]
dy = [1,-1,0,0,0,0]
dz = [0,0,0,0,1,-1]
day = 0
d_st = 0
count = 0
while q:
    z, y, x = q.popleft()
    for i in range(6):
        nx, ny, nz = x + dx[i], y + dy[i], z + dz[i]
        if 0 <= nx < M and 0 <= ny < N and 0 <= nz < H and box[nz][ny][nx] == 0:
            q.append((nz, ny, nx))
            box[nz][ny][nx] = 1
            count += 1
    d_st += 1
    if d_st == d_end:
        d_st = 0
        d_end = count
        count = 0
        day += 1
def check(day):
    for i in range(len(box)):
        for j in range(len(box[i])):
            for k in range(len(box[i][j])):
                if box[i][j][k] == 0:
                    return -1
    return day - 1
print(check(day))

File: 백준/Gold/7569. 토마토/토마토.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.*;

public class Main {
    static int[][][] box;
    static List<PointXYZ> first = new ArrayList<>();
    static int H, N, M;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        M = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());
        H = Integer.parseInt(st.nextToken());

        box = new int[H][N][M];

        for (int i = 0; i < H; i++) {
            for (int j = 0; j < N; j++) {
                st = new StringTokenizer(br.readLine(), " ");
                for (int k = 0; k < M; k++) {
                    box[i][j][k] = Integer.parseInt(st.nextToken());
                    if (box[i][j][k] == 1) {
                        first.add(new PointXYZ(i, j, k));
                    }
                }
            }
        }
        int day = bfs();

        bw.write(String.valueOf(day));

        br.close();

        bw.flush();
        bw.close();
    }

    private static int bfs() {
        Queue<PointXYZ> q = new LinkedList<>();
        int x, y, z, nx, ny, nz;
        int[] dx = {0,0,-1,1,0,0};
        int[] dy = {1,-1,0,0,0,0};
        int[] dz = {0,0,0,0,1,-1};
        int day = 0, d_st = 0, d_en = 0, count = 0;

        for (PointXYZ point : first) {
            q.offer(point);
            d_en++;
        }

        while (!q.isEmpty()) {
            PointXYZ point = q.poll();
            x = point.getX();
            y = point.getY();
            z = point.getZ();
            for (int i = 0; i < 6; i++) {
                nx = x + dx[i];
                ny = y + dy[i];
                nz = z + dz[i];

                if ((0 <= nx && nx < M) && (0 <= ny && ny < N) && (0 <= nz && nz < H) && box[nz][ny][nx] == 0) {
                    q.offer(new PointXYZ(nz, ny, nx));
                    box[nz][ny][nx] = 1;
                    count++;
                }
            }
            d_st++;

            if (d_st == d_en) {
                d_st = 0;
                d_en = count;
                count = 0;
                day++;
            }
        }

        for (int i = 0; i < H; i++) {
            for (int j = 0; j < N; j++) {
                for (int k = 0; k < M; k++) {
                    if (box[i][j][k] == 0) {
                        return -1;
                    }
                }
            }
        }

        return day - 1;
    }
}

class PointXYZ {
    private int X, Y, Z;
    public PointXYZ(int z, int y, int x) {
        this.Z = z;
        this.Y = y;
        this.X = x;
    }

    public int getX() {
        return this.X;
    }

    public int getY() {
        return Y;
    }

    public int getZ() {
        return Z;
    }
}